File: backend/db/models/camera.py
"""Camera model — registry semua kamera. Detail konfigurasi di config_json."""
from datetime import datetime, timezone
from sqlalchemy import String, Boolean, Integer, DateTime, JSON, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship
from backend.db.base import Base


class Camera(Base):
    __tablename__ = "cameras"
    id: Mapped[str] = mapped_column(String(20), primary_key=True)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    location: Mapped[str | None] = mapped_column(String(100))
    rtsp_main: Mapped[str] = mapped_column(String(500), nullable=False)
    rtsp_sub: Mapped[str | None] = mapped_column(String(500))

    # Dual Stream properties
    rtsp_url_main: Mapped[str | None] = mapped_column(String(500), nullable=True)
    rtsp_url_sub: Mapped[str | None] = mapped_column(String(500), nullable=True)

    storage_drive: Mapped[str] = mapped_column(String(50), nullable=False)
    motion_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    retention_days: Mapped[int] = mapped_column(Integer, default=30)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    config_json: Mapped[dict | None] = mapped_column(JSON)
    status: Mapped[str] = mapped_column(String(20), default="offline")

    # Recording schedule properties
    recording_schedule: Mapped[str] = mapped_column(String(20), default="24h", nullable=False)
    schedule_start_time: Mapped[str | None] = mapped_column(String(5), nullable=True)
    schedule_end_time: Mapped[str | None] = mapped_column(String(5), nullable=True)
    schedule_days: Mapped[str | None] = mapped_column(String(50), nullable=True)

    # Camera groups
    group_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("camera_groups.id", ondelete="SET NULL"), nullable=True)
    group: Mapped["CameraGroup | None"] = relationship(back_populates="cameras")

    last_seen: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))
    recordings: Mapped[list["Recording"]] = relationship(back_populates="camera", lazy="dynamic")
    motion_events: Mapped[list["MotionEvent"]] = relationship(back_populates="camera", lazy="dynamic")

File: mobile/lib/models/camera.dart
class Camera {
  final String id;
  final String name;
  final String? location;
  final String rtspMain;
  final String? rtspSub;
  final String storageDrive;
  final bool motionEnabled;
  final int retentionDays;
  final bool isActive;
  final bool isOnline;
  final int sortOrder;
  final DateTime createdAt;
  final DateTime updatedAt;

  Camera({
    required this.id,
    required this.name,
    this.location,
    required this.rtspMain,
    this.rtspSub,
    required this.storageDrive,
    required this.motionEnabled,
    required this.retentionDays,
    required this.isActive,
    required this.isOnline,
    required this.sortOrder,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Camera.fromJson(Map<String, dynamic> json) => Camera(
    id:             json['id'] as String,
    name:           json['name'] as String,
    location:       json['location'] as String?,
    rtspMain:       json['rtsp_main'] as String,
    rtspSub:        json['rtsp_sub'] as String?,
    storageDrive:   json['storage_drive'] as String,
    motionEnabled:  json['motion_enabled'] as bool? ?? false,
    retentionDays:  json['retention_days'] as int? ?? 30,
    isActive:       json['is_active'] as bool? ?? true,
    isOnline:       json['is_online'] as bool? ?? false,
    sortOrder:      json['sort_order'] as int? ?? 0,
    createdAt:      DateTime.parse(json['created_at'] as String),
    updatedAt:      DateTime.parse(json['updated_at'] as String),
  );

  Map<String, dynamic> toJson() => {
    'id': id, 'name': name, 'location': location,
    'rtsp_main': rtspMain, 'rtsp_sub': rtspSub,
    'storage_drive': storageDrive, 'motion_enabled': motionEnabled,
    'retention_days': retentionDays, 'is_active': isActive,
    'is_online': isOnline, 'sort_order': sortOrder,
    'created_at': createdAt.toIso8601String(),
    'updated_at': updatedAt.toIso8601String(),
  };
}
